core/state/statedb.go

```diff
@@ -154,25 +160,30 @@ func newStateDB(root common.Hash, db Database, snaps *snapshot.Tree) (*StateDB,
 		journal:             newJournal(),
 		hasher:              crypto.NewKeccakState(),
 	}
-	tr, err := db.OpenTrie(root)
-	if err != nil {
-		return nil, err
-	}
-	sdb.trie = tr
 	if sdb.snaps != nil {
 		if sdb.snap = sdb.snaps.Snapshot(root); sdb.snap != nil {
 			sdb.snapDestructs = make(map[common.Address]struct{})
 			sdb.snapAccounts = make(map[common.Address][]byte)
 			sdb.snapStorage = make(map[common.Address]map[string][]byte)
 		}
 	}
+
+	snapVerified := sdb.snap != nil && sdb.snap.Verified()
+	tr, err := db.OpenTrie(root)
+	// return error when 1. failed to open trie and 2. the snap is nil or the snap is not nil and done verification
+	if err != nil && (sdb.snap == nil || snapVerified) {
+		return nil, err
+	}
+	sdb.trie = tr
 	return sdb, nil
 }
 
 // StartPrefetcher initializes a new trie prefetcher to pull in nodes from the
 // state trie concurrently while the state is mutated so that when we reach the
 // commit phase, most of the needed data is already hot.
 func (s *StateDB) StartPrefetcher(namespace string) {
+	s.prefetcherLock.Lock()
+	defer s.prefetcherLock.Unlock()
```

Comment:
This PR adds a new goroutine to for pipeline commit?
Prefetcher will be called both on commit and verify routine now?
It is a bit complicated.

Reply:
```
"err != nil" means OpenTrie failed, newStateDB should fail as well?
newStateDB will success now, if OpenTrie fail when snap!=nil and snap verify failed.
It is make sense?
```
We allow `OpenTrie` failed for `newStateDB`, since Trie is not the necessary condition for block process.

Reply:
```
This PR adds a new goroutine to for pipeline commit?
Prefetcher will be called both on commit and verify routine now?
It is a bit complicated.
```
Prefetcher is only called when processing the block(apply transactions)